### PR TITLE
Webpack Helpers v1 Beta: Webpack 5, addFilter()

### PR DESCRIPTION
**Pull request by kadamwhite**
This PR changes a _few_ small things.... 😅 

- Update to support Webpack 5 and Webpack DevServer 4
- Rip out the filterLoaders system in favor of an `addFilter` more WP-style system I came up with while discussing an issue with @tfrommen
- Adjust or change a number of loaders and plugins for Webpack 5 compatibility
- Introduce a sourcemap loader and a few updated configuration options from the latest versions of CRA.

Known issues
- Production asset manifest is getting `auto` in the paths due to a `publicPath` change; need to check that.
- ESLint is now handled via a plugin, the loader is deprecated.

### CODE REVIEW

File: docs/changelog.md [hunk 1]
```diff
@@ -6,8 +6,15 @@ nav_order: 10
...
 
-## Next
-
+## v1.0
+
+- **Breaking**: End support for Node v10. Node v12.13 or later is now required.
+- **Breaking**: Switch to Webpack 5 and Webpack DevServer 4
+- **Breaking**: Replace `filterLoaders` system with [individual hooks accesible via the new `addFilter` helper](https://humanmade.github.io/webpack-helpers/modules/presets.html#customizing-presets).
+- **Potentially Breaking**: Remove `loaders.url()` and `loaders.file()` in favor of Webpack 5 [`asset` modules](https://webpack.js.org/guides/asset-modules/), now usable by including `loaders.asset()` (for assets which can be inlined) and `loaders.resource()` (as a catch-all for other types) in your module rules list. Asset modules are handled automatically in both presets, so this is only breaking if `loaders.url()` or `loaders.file()` was used directly.
+- **Potentially Breaking**: Remove `loaders.url()` in favor of Webpack 5 [`asset` modules](https://webpack.js.org/guides/asset-modules/), now usable by including `loaders.asset()` in your module rules list. Asset modules are handled automatically in both presets, so this is only breaking if `loaders.url()` was used directly.
```
Comment (tfrommen) [inline docs/changelog.md:15]: These seem to be partial duplicates?

File: src/loaders.js [hunk 2]
```diff
@@ -102,11 +108,21 @@ loaders.sass.defaults = {
...
+	loader: require.resolve( 'source-map-loader' ),
 	options: {},
 };
 
+loaders.resource.defaults = {
+	// Exclude `js` files to keep "css" loader working as it injects
+	// its runtime that would otherwise be processed through "file" loader.
+	// Also exclude `html` and `json` extensions so they get processed
+	// by webpacks internal loaders.
+	exclude: [ /^$/, /\.(js|mjs|jsx|ts|tsx)$/, /\.html$/, /\.json$/ ],
```
[the remark this answers is not in the document]
Comment (tfrommen) [reply src/loaders.js:124]: What about `.json5`? 🤔

**Review by tfrommen — COMMENTED** [review]
This is some great work! I'd like to test this on one or two projects, but here's some ad hoc feedback.

File: src/presets.js [hunk 3]
```diff
@@ -138,48 +98,51 @@ const development = ( config = {}, options = {} ) => {
...
-						getFilteredLoader( 'js' ),
-						// Convert small files to data URIs.
-						getFilteredLoader( 'url' ),
+						loaders.js(),
+						// Handle static asset files.
+						loaders.assets(),
 						// Parse styles using SASS, then PostCSS.
-						filterLoaders( {
+						// Permit filtering either on an env-specific or global basis.
+						applyFilters( 'preset/stylesheet-loaders', applyFilters( 'preset/dev/stylesheet-loaders', {
```
[the remark this answers is not in the document]
Comment (kadamwhite) [reply src/presets.js:122]: Great suggestion, I set this filter system up to account for that but didn't think to use it

File: src/loaders.js [hunk 4]
```diff
@@ -80,15 +85,16 @@ loaders.postcss.defaults = {
...
 	options: {
 		postcssOptions: {
-			plugins: [
+			ident: 'postcss',
+			plugins: applyFilters( 'loader/postcss/plugins', [
 				postcssFlexbugsFixes,
-				postcssPresetEnv( {
+				postcssPresetEnv( applyFilters( 'loader/postcss/preset-env', {
 					autoprefixer: {
 						flexbox: 'no-2009',
```
[the remark this answers is not in the document]
Comment (alwaysblank) [reply src/loaders.js:93]: IMO we shouldn't be trying to auto-prefix or auto-anything CSS grid--using grid should be a per-project decision that comes down to "do we need to support IE or can we live in the year 2022 with the rest of the world?" My feeling is that one of the big benefits of grid is its (relatively) simple API and what you can accomplish with that, and although you can support _some_ stuff in IE though the use of vendor prefixed properties, etc, it creates an unreasonable mental load for engineers just to support an ancient browser that constitutes a sliver of global usage. The limitations imposed by the subset of features that IE (kind of) supports make it difficult to really _use_ grid, and IME it's frequently just easier to use flexbox in that situation, as it's better supported and with some cleverness you can get a lot of the same layout effects.

tl;dr:

- Supporting grid in IE 11 is not easily automateable
- IE has extremely low usage
- We should default to "IE is not supported"
- If IE support is a project requirement, then we should just forgo grid on that project
- Therefore we don't need autoprefixer to try and do anything with grid

[the remark this answers is not in the document]
Comment (kadamwhite) [reply src/loaders.js:93]: I'm inclined to agree with Ben because that feels both well-argued, and in line with what other boilerplates and standardized build processes do. Thank you for the opinion and the argument!

**Comment by Sephsekla** [comment]
This is really awesome! Just gave it a quick "update and see what happens" test locally on a project.

### Build

Using the production preset, things were pretty smooth - one deprecation nag around a destructured import, and I needed to specifically install `prop-types`, which wasn't an explicit dependency before. After a couple of 10 second fixes, I'm seeing substantial speed improvements.

### Dev Server

DevServer seems to have more breaking changes, at least for our setup. I'll have to look into it further when I have time, but our array of entries mapped to devServers are no longer happy to use the same port: `Error: Unique ports must be specified for each devServer option in your webpack configuration.` I've seen this pattern on a couple of projects, so would be good to figure out a rock solid replacement.

Relevant part of the config is here

<details>
<summary><strong>devServer</strong></summary>

```js
	return choosePort( 9090 ).then( ( port ) => entries.map( ( entry ) => {
		return presets.development( {
			...common,
			devServer: {
				...devServer,
				port,
			},
			...entry,
			output: {
				...entry.output,
				publicPath: `https://${ host }:${ port }/${ entry.name.toLowerCase() }/`,
			},
		} );
	} ) );
};
```
</details>

where each item in `entries` looks something like 

<details>
<summary><strong>entrypoints</strong></summary>

```js
const example = {
	name: 'example',
	entry: {
		'lorem': filePath( 'content/plugins/example/assets/src/index.js' ),
	},
	output: {
		path: filePath( 'content/plugins/example/assets/build' ),
	},
};
```
</details>

Something like this *seems* to work, with a couple of angry messages about running webpack twice for some entrypoints, but I haven't yet had a chance to dig into it too much.

<details>
<summary><strong>potential devServer</strong></summary>

```js
return entries.map( ( entry, index ) =>
		choosePort( 9090 + index ).then( ( port ) => presets.development( {
			...common,
			devServer: {
				...devServer,
				port,
			},
			...entry,
			output: {
				...entry.output,
				publicPath: `https://${ host }:${ port }/${ entry.name.toLowerCase() }/`,
			},
		} ) )
	);
```
</details>

**Comment by kadamwhite** [comment]
@Sephsekla This is fantastic feedback, thank you! Making a reminder to myself to come back to this when we're at work next week to discuss what sort of dev server migration docs would be best.

**Comment by kucrut** [comment]
I couldn't get the dev server working with multiple configs on this new version. Webpack is complaining about being ran multiple times because the config that was passed to `webpack-dev-server` is an array of configs.

This time I'm doing it a bit differently. Instead of passing config to `webpack-dev-server`, the dev server is started manually via a node script. So far so good, apart from the HMR not working (yet). The dev server is running on `http://localhost` even though the project itself is running on https. This should be fine since localhost is special.

<details>
<summary><strong>Development Config</strong></summary>

```js
const { helpers, plugins, presets } = require( '@humanmade/webpack-helpers' );
const webpack = require( 'webpack' );
const WebpackDevServer = require( 'webpack-dev-server' );

const entries = require( './entries' );

const { choosePort, filePath } = helpers;

choosePort( 9090 ).then( async port => {
	const all = entries.map( config => {
		const outputPath = config.output.path.replace( filePath(), '' );

		config.output.publicPath = `http://localhost:${ port }${ outputPath }/`;

		config.plugins = [
			...( config.plugins || [] ),
			plugins.clean( {
				verbose: true,
				cleanOnceBeforeBuildPatterns: [ '**/*', '!asset-manifest.json' ],
			} ),
			plugins.fixStyleOnlyEntries(),
			plugins.manifest( {
				fileName: 'asset-manifest.json',
				writeToFileEmit: true,
			} ),
		];

		return presets.development( config );
	} );

	const server = new WebpackDevServer(
		{
			port,
			allowedHosts: 'all',
			client: {
				logging: 'error',
				webSocketURL: `ws://localhost:${ port }/ws`,
			},
			compress: true,
			headers: {
				'Access-Control-Allow-Origin': '*',
				'Access-Control-Allow-Methods': '*',
				'Access-Control-Allow-Headers': '*',
			},
			historyApiFallback: false,
			host: '0.0.0.0',
			hot: 'only',
		},
		webpack( all ),
	);

	await server.start();
} );
```
</details>

File: docs/guides/upgrading-to-v1.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+## Breaking Changes
+
+### Handling multi-config arrays in DevServer
+
+In a large project you may have a Webpack configuration that [exports multiple separate Webpack configuration objects](https://webpack.js.org/configuration/configuration-types/#exporting-multiple-configurations) in a multi-config array. With Webpack 4 and the pre-1.0 version of this helper library, you may have code which maps over these arrays and adds a `devServer` property to each one:
+
+```js
+```
+
+Previously, if you exported an array of configurations from your Webpack development config, each one could have an identical `devServer` property and Webpack DevServer would handle all configs with the same server. This let us
```
Comment (tfrommen) [inline docs/guides/upgrading-to-v1.md:26]: Just noting that this seems unfinished. Good cliffhanger, though. 😅

Comment (kadamwhite) [reply docs/guides/upgrading-to-v1.md:26]: Gotta make sure they're excited to buy the next upgrade! 😓 May not have time to flesh this out for a bit, would love help (from you or anybody reading) in fleshing out the upgrade guide.

**Comment by kadamwhite** [comment]
Summing up a long thread in Slack with @kucrut that we're running into this: https://github.com/webpack/webpack-dev-server/issues/2792 when using multiple entries in one config. `dependOn` works, in Dzikri's testing `runtimeChunk` didn't work as advertised in that thread. We may also want to drop the hashes from dev config entries in dev server mode. We may _also_ want to evaluate whether we should set a `host` value on the devServer object when we are transforming that in the dev preset.

**Comment by kadamwhite** [comment]
@tfrommen I don't think we've made any changes which would break that usage. The intent is absolutely to support that.

**Comment by mattheu** [comment]
I see there has been some discussion here on some of the issues i've run into...  and maybe some fixes too. But I'll note my workarounds in case they're helpful. 

* Multiple webpack configs are not supported. But if you name them e.g. `theme`, you can run an individual one like  `--config-name=theme`. I set up some aliases in my package.json scripts like `npm run start:theme`. There is a bit more overhead for the engineer in this, but for our project this was more performant and typically you don't need to run them both at once. 
* Single config but multiple JS files emitted. Need to set in webpack config `optimization: { runtimeChunk: 'single' }` and then ensure this is loaded once on the page, and set as a dependency of any JS files I'm loading. I have a custom asset loader function that wraps the HM Asset_Loader library to ensure this is done.

**Comment by kadamwhite** [comment]
> Multiple webpack configs are not supported. But if you name them e.g. theme, you can run an individual one like --config-name=theme. I set up some aliases in my package.json scripts like npm run start:theme. There is a bit more overhead for the engineer in this, but for our project this was more performant and typically you don't need to run them both at once.

This works well, and I always recommend labeling each config with a `name` so that you can do things this way. The issue with that approach happens when multiple configurations build to the same directory. We support doing so, with a shared manifest (#154), but imagine this scenario:

- Configuration `editor` builds these bundles:
  - `editor.js`
  - `editor.css`
- Configuration `frontend` builds these bundles:
  - `frontend.js`
  - `frontend.css`
  - `big-script-for-that-one-complex-page.js`

The production manifest would look like this:

```json
{
  "editor.js": "./editor.hash.js",
  "editor.css": "./editor.hash.css",
  "frontend.js": "./frontend.hash.js",
  "frontend.css": "./frontend.hash.css",
  "big-script-for-that-one-complex-page.js": "./big-script-for-that-one-complex-page.hash.js"
}
```

If you then run `npm run dev -- --config-name=editor`, you'll end up with a development manifest like this:

```json
{
  "editor.js": "http://localhost:8080/editor.js"
}
```

The way asset loader works now, if you use 
```php
$active_manifest = get_active_manifest( [ 'dev-manifest.json', 'prod-manifest.json' ] );

Asset_Loader\enqueue_asset( $active_manifest, 'editor.js' );

Asset_Loader\enqueue_asset( $active_manifest, 'frontend.js' );
```
then `editor.js` will get loaded, but `frontend.js` will fail (silently!) because it isn't present in the dev manifest.

If we use `--config-name` heavily, given that we do have a number of projects that build multiple configurations to a single folder, we'd probably need to also adjust the `enqueue_asset` function to look through each manifest in turn and enqueue the first located version of the requested asset, so that you could _fall back_ to the prod build if the script you need to load isn't available in the dev server.

**Comment by mattheu** [comment]
Something that was asked on a call. If you have to run multiple configs individually, can you just open up a new terminal tab and run the other config over there? And the answer is _kind of_. It works but hot module reloading fails for the second JS file loaded. I tried loading both only a single runtime JS file, and also multiple but no luck. Maybe there is a workaround?

**Comment by kadamwhite** [comment]
Superseded by #218, will close following a check to see if we want to bring over the new filter API
